cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/upgrade/TemplateSetUpgrader.java

```diff
@@ -100,30 +97,29 @@ public TemplateSetUpgrader() {
   public Map<com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration, Path> upgradeTemplatesToTemplateSets(
       Path contextLocation) throws Exception {
 
-    Path cobigenDir = contextLocation;
-    while (!cobigenDir.endsWith(ConfigurationConstants.COBIGEN_CONFIG_FILE)) {
-      cobigenDir = cobigenDir.getParent();
-      if (cobigenDir == null) {
-        cobigenDir = CobiGenPaths.getCobiGenHomePath();
-        cobigenDir.resolve(ConfigurationConstants.COBIGEN_CONFIG_FILE);
-      }
-    }
-    ContextConfiguration contextConfiguration = getContextConfiguration(contextLocation);
+    Path cobigenHome = CobiGenPaths.checkCustomHomePath(contextLocation);
+    if (cobigenHome == null)
+      cobigenHome = CobiGenPaths.getCobiGenHomePath();
 
-    Path templateSets = Files.createDirectory(cobigenDir.resolve(ConfigurationConstants.TEMPLATE_SETS_FOLDER));
-    Path adapted = Files.createDirectory(templateSets.resolve(ConfigurationConstants.ADAPTED_FOLDER));
-    Path cobigenTemplates = cobigenDir.resolve(ConfigurationConstants.TEMPLATES_FOLDER)
-        .resolve(ConfigurationConstants.COBIGEN_TEMPLATES);
-    Path templates = cobigenTemplates.resolve(ConfigurationConstants.TEMPLATE_RESOURCE_FOLDER);
+    List<Path> newContextLocation = FileSystemUtil.collectAllContextXML(contextLocation);
+    Path folderOfContextLocation = newContextLocation.get(0).getParent();
+    ContextConfiguration contextConfiguration = getContextConfiguration(folderOfContextLocation);
+    Path templateSets = cobigenHome.resolve(ConfigurationConstants.TEMPLATE_SETS_FOLDER);
+    if (!Files.exists(templateSets))
+      Files.createDirectory(templateSets);
+    Path adapted = templateSets.resolve(ConfigurationConstants.ADAPTED_FOLDER);
+    if (!Files.exists(adapted))
+      Files.createDirectory(adapted);
+    Path cobigenTemplatesFolder = CobiGenPaths.getTemplatesPomFileLocation(newContextLocation.get(0));
 
     List<com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration> contextFiles = splitContext(
         contextConfiguration);
     Map<com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration, Path> contextMap = new HashMap<>();
     for (com.devonfw.cobigen.impl.config.entity.io.v3_0.ContextConfiguration cc : contextFiles) {
       for (com.devonfw.cobigen.impl.config.entity.io.v3_0.Trigger trigger : cc.getTrigger()) {
-        Path triggerFolder = templates.resolve(trigger.getTemplateFolder());
+        Path triggerFolder = folderOfContextLocation.resolve(trigger.getTemplateFolder());
         Path newTriggerFolder = adapted.resolve(trigger.getTemplateFolder());
-        Path utilsPath = cobigenTemplates.resolve(ConfigurationConstants.UTIL_RESOURCE_FOLDER);
+        Path utilsPath = folderOfContextLocation.getParent().resolve("java");
```

Comment:
What happens if this path is null?

Reply:
I added a check when copying the folder later at line 132.